genkit-tools/cli/src/utils/updates.ts

```diff
@@ -0,0 +1,246 @@
+    if (error instanceof GenkitToolsError) {
+      throw error;
+    }
+
+    throw new GenkitToolsError(
+      `Failed to fetch npm versions: ${error?.message ?? error}`
+    );
+  }
+}
+
+/**
+ * Checks if update notifications are disabled via environment variable or user config.
+ */
+function areUpdateNotificationsDisabled(): boolean {
+  if (process.env.GENKIT_CLI_DISABLE_UPDATE_NOTIFICATIONS === 'true') {
+    return true;
+  }
+  const userSettings = getUserSettings();
+  return Boolean(userSettings[UPDATE_NOTIFICATIONS_OPT_OUT_CONFIG_TAG]);
+}
+
+/**
+ * Gets the latest version and update message for compiled binary installations.
+ */
+async function getBinaryUpdateInfo(): Promise<{
+  latestVersion: string;
+  updateMessage: string;
+} | null> {
+  const gcsLatestData = await getGCSLatestData();
+  const machine = `${platform}-${arch}`;
+  const platformData = gcsLatestData.platforms[machine];
+
+  if (!platformData) {
+    logger.debug(`No update information for platform: ${machine}`);
+    return null;
+  }
+
+  const latestVersion = normalizeVersion(gcsLatestData.latestVersion);
+  const fileName = platformData.versionedUrl.split('/').pop() || '';
+  const downloadUrl = `${GCS_BUCKET_URL}/prod/${machine}/v${latestVersion}/${fileName}`;
+  const updateMessage = `${clc.dim('Run')} ${clc.bold(`curl -Lo ./genkit_bin ${downloadUrl}`)} ${clc.dim('to upgrade')}`;
+  return { latestVersion, updateMessage };
+}
+
+/**
+ * Gets the latest version and update message for npm installations.
+ */
+async function getNpmUpdateInfo(): Promise<{
+  latestVersion: string;
+  updateMessage: string;
+} | null> {
+  const latestVersion = await getLatestVersionFromNpm();
+  if (!latestVersion) {
+    logger.debug('No available versions found from npm.');
+    return null;
+  }
+  const updateMessage = `${clc.dim('Update')} ${clc.bold('genkit-cli')} ${clc.dim('using your preferred package manager (npm, pnpm, yarn, etc.)')}`;
+  return { latestVersion, updateMessage };
+}
+
+/**
+ * Shows an update notification if a new version is available.
+ * This function is designed to be called from the CLI entry point.
+ * It can be disabled by the user's configuration or environment variable.
+ */
+export async function showUpdateNotification(): Promise<void> {
+  try {
+    if (areUpdateNotificationsDisabled()) {
+      return;
+    }
+
+    const { isCompiledBinary } = detectCLIRuntime();
+    const updateInfo = isCompiledBinary
+      ? await getBinaryUpdateInfo()
+      : await getNpmUpdateInfo();
+
+    if (!updateInfo) {
+      return;
+    }
+
+    const { latestVersion, updateMessage } = updateInfo;
+    const current = normalizeVersion(currentVersion);
+
+    if (!semver.valid(latestVersion) || !semver.valid(current)) {
+      logger.debug(
+        `Invalid semver: current=${current}, latest=${latestVersion}`
+      );
+      return;
+    }
+
+    if (!semver.gt(latestVersion, current)) {
+      return;
+    }
+
+    logger.info(
```

Comment:
lets make the message look very similar to firebase cli:
https://github.com/firebase/firebase-tools/blob/0f06446371dcd02174f3ea23d7715de3a2b30cdf/src/bin/cli.ts#L80